### PR TITLE
WIP: Rework go builds to use a makefile and containerize the entire process.

### PR DESCRIPTION
**Pull request by spxtr**
This lets us do a release either entirely in a container or else not at all in a container. You can say `make quick-release-local -j4` and get an incremental build if you have set up your `GOPATH` properly, or you can say `make quick-release` and get a completely containerized, hermetic build. The incremental build still takes a non-zero amount of time because it needs to tar up all the things and build docker images and the like, but the go part of it is much faster.

On my workstation, `make quick-release-local -j4` takes ~5:30 the first time and ~1:50 the second. The old `make quick-release` takes ~6:30 the first time and  ~3:40 the second. The go tool is pretty good at parallelizing its build, so more than 8 concurrent jobs actually hurt performance.

I also made some rules for building kubernetes binaries, so now `make kubectl` or `make kube-apiserver` work nicely if your `GOPATH` is set.

There's still quite a bit of work to go before this is ready to go in, but I'd like to get it out there for comments. I still need to:
- update docs
- figure out what to do with the build-image tarball that I've commented out, get the source tar into the final tar
- delete a bunch of old shell code that is deactivated by this change
- make some of the `build/common.sh` code a little better at handling non-hermiticity
- make the verify checks just use the makefile
- make the things not owned by root (#25525)
- make it play better with ci

cc @thockin @ixdy @rmmh @david-mcmahon

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/27746)

### CODE REVIEW

**Comment by thockin-cc** [comment]
OMG.  Please wait.  I have a MEGA pull request that touches a lot of the
same files.  I agree with the goal, but I am begging that we spend time
talking before we diverge any further.

**Comment by spxtr** [comment]
SGTM. I'm not in a rush.

**Comment by spxtr** [comment]
@thockin You're referring to the generated code PR (#25978), right? I'm happy to wait until after that is in to move this forward.

File: test/e2e/scheduler_predicates.go [hunk 1]
```diff
@@ -45,7 +45,7 @@ var masterNodes sets.String
 func getPodsScheduled(pods *api.PodList) (scheduledPods, notScheduledPods []api.Pod) {
 	for _, pod := range pods.Items {
 		if !masterNodes.Has(pod.Spec.NodeName) {
-			if pod.Spec.NodeName != "" {
+			if pod.Spec.NodeName != "asdf" {
```
Comment (rmmh) [inline test/e2e/scheduler_predicates.go]: ?

Comment (spxtr) [inline test/e2e/scheduler_predicates.go]: Whoops. That was to check that I was using teststale properly :)

File: Makefile [hunk 2]
```diff
@@ -52,7 +52,7 @@ export KUBE_GOLDFLAGS
 # Example:
 #   make release
 all release:
-	@docker run --rm -it \
+	@docker run --rm -i \
```
Comment (rmmh) [inline Makefile]: this turns *off* the tty

Comment (spxtr) [inline Makefile]: Hah, trust the code, not the commit message.

**Comment by spxtr** [comment]
Ooh interesting. @fejta, on an `n1-standard-8` I had no issues, but on an `n1-highcpu-16` machine I got,
```
/usr/local/go/pkg/tool/linux_amd64/link: running cc failed: fork/exec /usr/bin/cc: cannot allocate memory
```

**Comment by thockin-cc** [comment]
Yes.  I am at 30+ commits so far, and have revised countless times :)

File: test/e2e/scheduler_predicates.go [hunk 3]
```diff
@@ -45,7 +45,7 @@ var masterNodes sets.String
 func getPodsScheduled(pods *api.PodList) (scheduledPods, notScheduledPods []api.Pod) {
 	for _, pod := range pods.Items {
 		if !masterNodes.Has(pod.Spec.NodeName) {
-			if pod.Spec.NodeName != "asdf" {
```
Comment (david-mcmahon) [inline test/e2e/scheduler_predicates.go]: I suppose you'll squash this?

**Comment by david-mcmahon** [comment]
@spxtr This is great.  Thanks for doing this!

File: test/e2e/scheduler_predicates.go (hunk 3 above)
Comment (spxtr) [inline test/e2e/scheduler_predicates.go]: Will do once it's ready to go in.

**Comment by spxtr** [comment]
@thockin looks like the generated stuff has settled down. Is now a good time for me to move this forward?

**Comment by fejta** [comment]
@thockin please assign back to me when Joe can continue working on this

**Comment by spxtr** [comment]
I'll reopen a new PR once I'm ready to work on this again.

**Pull request closed without merging**